dnf/cli/commands/history.py

```diff
@@ -219,24 +243,140 @@ def str2transaction_id(s):
 
+        self._na_reason_cache = {}
+
+        for na, pkg_data in rpms.items():
+            subj = hawkey.Subject(na)
+            nas = subj.get_nevra_possibilities(forms=[hawkey.FORM_NA])
+
+            if len(nas) != 1:
+                raise dnf.exceptions.Error(_('Cannot parse name.arch for package "{na}".').format(na=na))
+
+            parsed_na = nas[0]
+            try:
+                action = pkg_data["action"]
+
+                # validate the reasons before running the transaction and cache them
+                try:
+                    self._na_reason_cache[na] = \
+                        libdnf.transaction.StringToTransactionItemReason(pkg_data["reason"])
+                except IndexError as e:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package reason "{reason}" for package "{pkg}" in "{fn}".')
+                            .format(reason=pkg_data["reason"], pkg=na, fn=fn)
+                    )
+
+                if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["to"], parsed_na.arch)
+                    if action in ("Upgrade", "Downgrade", "Reinstall"):
+                        from_nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                        query = dnf.subject.Subject(from_nevra).get_best_solution(self.base.sack)["query"]
+                        if not query.installed():
+                            raise dnf.exceptions.Error(
+                                _('Package "{pkg}" not found installed for action "{action}" in "{fn}".')
+                                    .format(pkg=from_nevra, action=action, fn=fn)
+                            )
+
+                elif action == "Removed":
+                    nevra = "%s-%s.%s" % (parsed_na.name, pkg_data["from"], parsed_na.arch)
+                else:
+                    raise dnf.exceptions.Error(
+                        _('Unexpected value of package action "{action}" for package "{pkg}" in "{fn}".')
+                            .format(action=action, pkg=na, fn=fn)
+                    )
+
+                query = dnf.subject.Subject(nevra).get_best_solution(self.base.sack)["query"]
+
+                if not query:
+                    raise dnf.exceptions.Error(_('Package "{nevra}" from "{fn}" was not found.')
+                        .format(nevra=nevra, fn=fn))
+
+                if action in ("Install", "Upgrade", "Downgrade"):
+                    sltr = dnf.selector.Selector(self.base.sack).set(pkg=query)
+                    self.base.goal.install(select=sltr, optional=False)
+                elif action == "Reinstall":
+                    sltr = dnf.selector.Selector(self.base.sack).set(pkg=query.available())
+                    self.base.goal.install(select=sltr, optional=False)
+                elif action == "Removed":
+                    self.base.goal.erase(query.installed()[0])
```

Comment:
For replay, I would recommend to disable removal of unused dependencies.

Reply:
In my tests it didn't remove them... any tip on how should I do it?

Reply:
The setting can be adjusted by 'base.conf.clean_requirements_on_remove'.

Reply:
Yes. I guess it's one of the config options we should store into the transaction file.

I've just realized though these options aren't stored in the transaction history, so I can record what's in the config at the time of storing the transaction, but that can be different from what was in use when the transaction happened :slightly_frowning_face: So that's quite a big issue...

Reply:
You don't have to store it, you have to override it to false! Same for weak-deps.

Reply:
I guess I should, for `clean_requirements_on_remove`. `install_weak_deps` should produce the same results if I store it, so I'm not entirely sure which is better for that.

But the problem remains for the rest of the configuration that needs to be stored.

Reply:
@j-mracek I've originally compiled this list of configuration options that I wanted to store in the transaction file (`allow_erasing` is a demand, not a config option):
```
allow_erasing
best
clean_requirements_on_remove
ignorearch
install_weak_deps
multilib_policy
obsoletes
```

Going through the list, I _think_ that the only configuration option that will really have any effect is `ignorearch` (after I set `clean_requirements_on_remove` and `install_weak_deps` to false while replaying). Reason being the transaction already contains the full list of packages, so if I'm not mistaken on something, the options don't actually make a difference.

With the exception of `ignorearch`, which I think is quite specialized and has a big effect on the system. Given the fact that I can't store the configuration from the time the transaction was run, I'm seriously considering just leaving the whole feature (of storing the configuration in the transaction file) out, at least for now - until we have a good reason to add it. What do you think?

Reply:
I also suggest that there is no requirement to store any additional configurations.